hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieDefaultTimeline.java

```diff
@@ -95,6 +95,13 @@ public HoodieTimeline filterPendingExcludingCompaction() {
             && (!instant.getAction().equals(HoodieTimeline.COMPACTION_ACTION))), details);
   }
 
+  @Override
+  public HoodieTimeline filterPendingExcludingCompactionAndClustering() {
+    return new HoodieDefaultTimeline(instants.stream().filter(instant -> (!instant.isCompleted())
+        && (!instant.getAction().equals(HoodieTimeline.COMPACTION_ACTION))
+        && (!instant.getAction().equals(HoodieTimeline.REPLACE_COMMIT_ACTION))), details);
```

Comment:
hmm, this is actually tricky. what if the replace_commit belongs to insert_overwrite? We have to rollback insert_overwrite related replacecommits, but not clustering related ones. 

We could move this method out of timeline into a utils method and parse content to filter only clustering instants.

Reply:
thanks, make sense .